### PR TITLE
Port more array-related derivatives from Nabla

### PR DESCRIPTION
**Pull request by ararslan**
The stuff in Nabla's `src/sensitivities/array.jl`, which includes `reshape`, `hcat`, `vcat`, and `fill`.

### CODE REVIEW

File: src/rules/array.jl [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+#####
+
+function rrule(::typeof(hcat), A::AbstractArray, Bs::AbstractArray...)
+    Y = hcat(A, Bs...)
+    Xs = (A, Bs...)
+    rules = ntuple(length(Bs) + 1) do i
+        l = mapreduce(j->size(Xs[j], 2), Base.add_sum, 1:i-1; init=0)
+        u = l + size(Xs[i], 2)
+        dim = u > l + 1 ? (l+1:u) : u
+        Rule(Ȳ->copy(selectdim(Ȳ, 2, dim)))
```
Comment (oxinabox) [inline src/rules/array.jl]: why `copy`?
For my education on how ChainRules thinks of defensive copies

File: src/rules/array.jl [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+#####
+##### `fill`
+#####
+
+function rrule(::typeof(fill), value::Any, dims::Tuple{Vararg{Int}})
+    return fill(value, dims), (Rule(sum), DNERule())
+end
+
+function rrule(::typeof(fill), value::Any, dims::Int...)
+    return fill(value, dims), (Rule(sum), fill(DNERule(), length(dims))...)
```
Comment (oxinabox) [inline src/rules/array.jl]:
```suggestion
    return fill(value, dims), (Rule(sum), Fill(DNERule(), length(dims))...)
```

I thought we added a FillArrays dependency?
He haven't but we should?

Comment (willtebbutt) [reply src/rules/array.jl]: I wouldn't be pro- introducing this here until we've addressed #46 . Also, it doesn't look like `Fill`s have been optimised for splatting into a `Tuple` yet, so there's no real performance gain here.

Comment (oxinabox) [reply src/rules/array.jl]: Oh, I missed that it was splatted
Fair enough

Comment (oxinabox) [reply src/rules/array.jl]: We can make this compile time, if we use `ntuple`
```
julia> @btime (()->tuple(ntuple(_->Nothing(), 10)...))()
  0.027 ns (0 allocations: 0 bytes)
(nothing, nothing, nothing, nothing, nothing, nothing, nothing, nothing, nothing, nothing)

julia> @btime (()->tuple(fill(Nothing(), 10)...))()
  319.146 ns (1 allocation: 80 bytes)
(nothing, nothing, nothing, nothing, nothing, nothing, nothing, nothing, nothing, nothing)
```


```suggestion 
    return fill(value, dims), (Rule(sum), ntuple(_->DNERule(), length(dims))...)
```

File: src/rules/array.jl (hunk 1 above)
Comment (ararslan) [reply src/rules/array.jl]: Definitely a defensive copy in this case; `selectdim` returns a view, which `copy` materializes. This is the same thing that Nabla does.

Comment (oxinabox) [reply src/rules/array.jl]: Right, I see why that is needed.

Thanks

Comment (ararslan) [reply src/rules/array.jl]: I should add a comment to that effect, so thanks for asking. :slightly_smiling_face:

**Pull request merged**